## TOPSIS (Technique for Order of Preference by Similarity to Ideal Solution)

|               | Price  | Storage Space (GB)| Camera (MP)| Looks |
|---------------|--------|-------------------|------------|-------|
| Phone 1       |  250   |   16              |  12        |   5   |
| Phone 2       |  200   |   16              |  8         |   3   |
| Phone 3       |  300   |   32              |  16        |   4   |
| Phone 4       |  275   |   32              |  8         |   4   |
| Phone 5       |  225   |   16              |  16        |   2   |



In [1]:
num_criteria = 4
prices = [ 250, 200, 300, 275, 225]
storage = [ 16, 16, 32, 32, 16]
camera = [12, 8, 16, 8, 16 ]
looks = [5, 3, 4, 4, 2]

In [2]:
import math

\\[ \bar X_{ij} = \frac{X_{ij}}{\sqrt{\sum_{j=1}^{n} X^2_{ij}}} \\]

In [3]:


def normalize_scale( vec, num_criteria):
    vec_squared = [ v**2 for v in vec]
    summation = sum(vec_squared)
    denom = math.sqrt(summation)
    return [v/denom/num_criteria for v in vec]

In [4]:
prices_n, storage_n, camera_n, looks_n = normalize_scale(prices, 4), normalize_scale(storage, 4), normalize_scale(camera,4), normalize_scale(looks, 4)


In [5]:
prices_n, storage_n, camera_n, looks_n

([0.11070186069251191,
  0.08856148855400953,
  0.1328422328310143,
  0.1217720467617631,
  0.09963167462326072],
 [0.07537783614444091,
  0.07537783614444091,
  0.15075567228888181,
  0.15075567228888181,
  0.07537783614444091],
 [0.10714285714285714,
  0.07142857142857142,
  0.14285714285714285,
  0.07142857142857142,
  0.14285714285714285],
 [0.1494035761667992,
  0.08964214570007951,
  0.11952286093343936,
  0.11952286093343936,
  0.05976143046671968])

$ V^{+}_{j} $ is the idea best value and $ V^{-}_{j} $ is the idea worst value

$ V^+_0 = 0.0885614 $ best_price is the lowest price. 

$ V^+_1 = 0.1508 $ best storage is the maximum storage



In [6]:
ideal_vector = [ min(prices_n), max(storage_n), max(camera_n), max(looks_n) ]
ideal_vector

[0.08856148855400953,
 0.15075567228888181,
 0.14285714285714285,
 0.1494035761667992]

In [7]:
anti_vector = [ max(prices_n), min(storage_n), min(camera_n), min(looks_n) ]
anti_vector

[0.1328422328310143,
 0.07537783614444091,
 0.07142857142857142,
 0.05976143046671968]

Euclidean distance from ideal best 
$$ S^{+}_i = \sqrt{ \sum_{j=1}^m \left ( V_{ij} - V^{+}_j \right )^2 } $$

Euclidean distance from ideal worst
$$ S^{-}_i = \sqrt{ \sum_{j=1}^m \left ( V_{ij} - V^{-}_j \right )^2 } $$

Performance Score $P_i$ is given by $$ P_i = \frac{S^{-}_i}{S^{+}_i + S^{-}_i} $$

In [8]:
options = []
options_n = []
for i in range(len(prices)):
    option = [ prices[i], storage[i], camera[i], looks[i]]
    option_n = [prices_n[i], storage_n[i], camera_n[i], looks_n[i]]
    options.append(option)
    options_n.append(option_n)

options_n

[[0.11070186069251191,
  0.07537783614444091,
  0.10714285714285714,
  0.1494035761667992],
 [0.08856148855400953,
  0.07537783614444091,
  0.07142857142857142,
  0.08964214570007951],
 [0.1328422328310143,
  0.15075567228888181,
  0.14285714285714285,
  0.11952286093343936],
 [0.1217720467617631,
  0.15075567228888181,
  0.07142857142857142,
  0.11952286093343936],
 [0.09963167462326072,
  0.07537783614444091,
  0.14285714285714285,
  0.05976143046671968]]

In [9]:
def distance( a, b):
    sd = 0.0
    for j in range(len(a)):
        sd += (a[j] - b[j])**2
    dist = math.sqrt( sd)
    return dist

diff_from_ideal = []
diff_from_worst = []
for option_n in options_n:
    d = distance( ideal_vector, option_n)
    d2 = distance( anti_vector, option_n)
    diff_from_ideal.append(d)
    diff_from_worst.append(d2)


In [10]:
diff_from_ideal

[0.08629904092358841,
 0.11981355336343749,
 0.053419485738657506,
 0.08424867438514544,
 0.11764387568904856]

In [11]:
diff_from_worst

[0.09900212405917001,
 0.053419485738657506,
 0.11981355336343749,
 0.09682869292133708,
 0.07877170807337568]

In [12]:
performance_score = []
for j in range(len(diff_from_ideal)):
    P = diff_from_worst[j]/ (diff_from_ideal[j] + diff_from_worst[j])
    performance_score.append(P)
performance_score

[0.5342768571821003,
 0.3083677687324685,
 0.6916322312675315,
 0.534736584486838,
 0.40104612151678615]